rules/integrations/aws/impact_s3_bucket_object_deletion_and_ransomware_note_added.toml

```diff
@@ -0,0 +1,125 @@
+
+### Additional Information:
+
+For further guidance on managing S3 bucket security and protecting against ransomware, refer to the [AWS S3 documentation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html) and AWS best practices for security. Additionally, consult the following resources for specific details on S3 ransomware protection:
+- [ERMETIC REPORT - AWS S3 Ransomware Exposure in the Wild](https://s3.amazonaws.com/bizzabo.file.upload/PtZzA0eFQwV2RA5ysNeo_ERMETIC%20REPORT%20-%20AWS%20S3%20Ransomware%20Exposure%20in%20the%20Wild.pdf)
+- [AWS S3 Ransomware Batch Deletion](https://stratus-red-team.cloud/attack-techniques/AWS/aws.impact.s3-ransomware-batch-deletion/)
+- [S3 Ransomware Part 1: Attack Vector](https://rhinosecuritylabs.com/aws/s3-ransomware-part-1-attack-vector/)
+
+"""
+references = [
+    "https://s3.amazonaws.com/bizzabo.file.upload/PtZzA0eFQwV2RA5ysNeo_ERMETIC%20REPORT%20-%20AWS%20S3%20Ransomware%20Exposure%20in%20the%20Wild.pdf",
+    "https://stratus-red-team.cloud/attack-techniques/AWS/aws.impact.s3-ransomware-batch-deletion/",
+    "https://rhinosecuritylabs.com/aws/s3-ransomware-part-1-attack-vector/",
+]
+risk_score = 73
+rule_id = "7fda9bb2-fd28-11ee-85f9-f661ea17fbce"
+setup = "AWS S3 data types need to be enabled in the CloudTrail trail configuration."
+severity = "high"
+tags = [
+    "Domain: Cloud",
+    "Data Source: AWS",
+    "Data Source: Amazon Web Services",
+    "Data Source: AWS S3",
+    "Use Case: Threat Detection",
+    "Tactic: Impact",
+]
+timestamp_override = "event.ingested"
+type = "eql"
+
+query = '''
+sequence by tls.client.server_name with maxspan=5m
+
+    /* sequence on tls.client.server_name is the s3 bucket name the client is accessing */
+    /* Check for multiple S3 bucket objects being accessed */
+    [any where event.dataset == "aws.cloudtrail"
+        and event.provider == "s3.amazonaws.com"
+        and event.action == "GetObject"
+        and event.outcome == "success"] with runs=5
+
+    /* Check for objects being deleted after accessed, indicating retrieval */
+    /* Ideally we would use the object key to correlate the objects being accessed and deleted */
+    [any where event.dataset == "aws.cloudtrail"
+        and event.provider == "s3.amazonaws.com"
+        and event.action in ("DeleteObjects", "DeleteObject")
+        and event.outcome == "success"]
+
+    /* Check for a ransom note being pushed to the same bucket objects were just retrieved and deleted from */
+    [any where event.dataset == "aws.cloudtrail"
+        and event.provider == "s3.amazonaws.com"
+        and event.action == "PutObject"
+        and event.outcome == "success"
+        and aws.cloudtrail.request_parameters regex ".*key=.*(txt|note|ransom|html|lock|crypt|enc|readme|how_to_decrypt|decrypt_instructions|recovery|datarescue)}"]
```

Comment:
Yes I agree it will be expensive for a couple reasons.

- 3 sequences to match on common API calls within a 5m maxspan, however, it is sequenced by a specific S3 bucket which helps
- the use of `any` does not help pre-filter on specific events, unfortunately there are no `event.category` fields for the data that comes from these specific service providers in AWS cloudtrail data, I've asked security integrations to update this. But it will require some specific additional considerations.
- We are unable to aggregate by specific object IDs in the S3 buckets for not only better performance but higher efficacy as a result of those being only available in `request_parameters` and `response_elements`

Couple of options:
- Remove the sequence, focus only on known S3 `PutObject` API calls where the extensions are known ransom notes
- Attempt to accomplish something similar with ES|QL where we can pre-process the data, normalize it a bit and do aggregations in one query


I'll place this back into draft and dig into it a bit more.

Reply:
@Aegrah - I have re-asked for your review. After further investigation and toying with different queries, I chose to limit this rule to `PutObject` API requests for files with common ransom note extensions. 

Some information to share - `aws.cloudtrail.request_parameters` is a keyword field and also has a related `aws.cloudtrail.flattened.request_parameters` field containing the JSON where keys can be used in KQL, but not in EQL unfortunately. If so we would have been able to sequence on S3 bucket name, ARN and the objects being retrieved and deleted. Ideally this would have been a decent rule to detect rapid retrieval and deletion which indicates theft and potential ransom activity.

Unfortunately ES|QL does not have the capability to store two tables and right outer join on existing fields or else we would be able to correlate GET and DELETE API calls accordingly as well.

Ultimately, it made more sense to focus on the ransom note upload and limit the aggregation stats to `1` because text files are commonly used for ransom notes, but also very common in S3 buckets.